src/sinks/util/encoding/mod.rs

```diff
@@ -106,12 +107,54 @@ pub trait Encoder<T> {
     }
 }
 
-impl<E, T> Encoder<T> for Arc<E>
-where
-    E: Encoder<T>,
-{
-    fn encode_input(&self, input: T, writer: &mut dyn io::Write) -> io::Result<usize> {
-        (**self).encode_input(input, writer)
+impl Encoder<Vec<Event>> for (Transformer, crate::codecs::Encoder<Framer>) {
+    fn encode_input(
+        &self,
+        mut events: Vec<Event>,
+        writer: &mut dyn io::Write,
+    ) -> io::Result<usize> {
+        let mut encoder = self.1.clone();
+        let num_events = events.len();
+        let mut bytes_written = 0;
+        match num_events {
+            0 => {
+                bytes_written += writer.write(encoder.batch_prefix())?;
+                bytes_written += writer.write(encoder.batch_suffix())?;
+            }
+            1 => {
+                let mut event = events.pop().unwrap();
+                bytes_written += writer.write(encoder.batch_prefix())?;
+                self.0.transform(&mut event);
+                let mut bytes = BytesMut::new();
+                encoder
+                    .serialize(event, &mut bytes)
+                    .map_err(|error| io::Error::new(io::ErrorKind::InvalidData, error))?;
+                bytes_written += writer.write(&bytes)?;
+                bytes_written += writer.write(encoder.batch_suffix())?;
+            }
```

Comment:
Is this variant needed? It seems that in the below variant, if you pop last, and the collection is empty, no iteration happens, and all the other code is the same as this variant.

Or is there a measurable performance improvement we want to capture here?

Reply:
Hm, true. That's a nice observation 😄

I don't think that invoking he empty iterator will have any significant impact, and since we're batching, the `1` case should be rare anyway

Reply:
I now remember that I had done that with correctness in mind, not performance. I previously had some code that would handle the first element differently, such that "one single element" and "multiple elements" had different logic.

You're right that it can be collapsed now, that is pretty neat!

Reply:
Applied that suggestion in 3221c9de0725a69fd60fa940a2dc5a09ab741458